text/3414-deprecate-spooky-dropck-at-a-distance.md

```diff
@@ -0,0 +1,362 @@
+outlive requirements, thus also exhibiting "spooky-dropck-at-a-distance".
+
+Simply defining every non-`needs_drop` type as being pure w.r.t. drop would,
+however, break `#[may_dangle]`, so we need to adapt it.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Changes to dropck (stable)
+
+Types like `fn(T)`, `ManuallyDrop<T>`, `PhantomData<T>`, `[T; 0]` and `&'a T`,
+that don't need to be dropped, place no liveness requirements on `'a` or `T`
+when they go out of scope - even if `T` would otherwise introduce liveness
+requirements.
+
+In other words, this compiles:
+
+```rust
+fn main() {
+    let mut x;
+    {
+        x = &String::new();
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+But a type which does need to be dropped, introduces liveness requirements:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x;
+    {
+        x = PrintOnDrop(&*String::new());
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+    // ERROR: temporary may not live long enough
+}
+```
+
+Unless it's in one of the above types:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x;
+    {
+        x = &PrintOnDrop(&*String::new());
+        // PrintOnDrop implicitly dropped here
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+As a special case, some built-in types, like `Vec`, `Box`, `Rc`, `Arc`,
+`HashMap`, among others, despite needing to be dropped, do not place liveness
+requirements unless `T` demands liveness requirements. This is okay:
+
+```rust
+fn main() {
+    let mut x = vec![];
+    {
+        x.push(&String::new());
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+But this isn't:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x = vec![];
+    {
+        x.push(PrintOnDrop(&*String::new()));
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+    // ERROR: temporary may not live long enough
+}
+```
+
+## Changes to `#[may_dangle]` (unstable)
+
+A type marked `#[may_dangle(drop)]` gets checked for liveness at drop. This is
+necessary for `Vec`:
+
+```rust
+struct Vec<T> {
+  ...
+}
+
+unsafe impl<#[may_dangle(drop)] T> Drop for Vec<T> {
+  fn drop(&mut self) {
+    ...
+  }
+}
+```
+
+So that this compiles:
+
+```rust
+fn main() {
+  let mut v = vec![];
+  {
+    v.push(&String::from("temporary"));
+  }
+}
+```
+
+But this cannot compile, as it would be unsound:
```

Comment:
(as above, clarify what has changed in this example such that it cannot compile; also, clarify why is it unsound)